app/modules/collaborations/models.py

```diff
@@ -221,11 +232,47 @@ def notify_pending_users(self):
                     )
 
     def _notify_user(self, sending_user_assoc, receiving_user_assoc, notification_type):
-        from app.modules.notifications.models import Notification, NotificationBuilder
+        from app.modules.notifications.models import (
+            Notification,
+            NotificationBuilder,
+            NotificationType,
+        )
 
         builder = NotificationBuilder(sending_user_assoc.user)
         builder.set_collaboration(self)
-        Notification.create(notification_type, receiving_user_assoc.user, builder)
+        notif = Notification.create(notification_type, receiving_user_assoc.user, builder)
+        # I expect this to be set by the builder but it sometimes does not, unsure why.
+        notif.sender_guid = sending_user_assoc.user.guid
+
+        # in these states, every notification is considered to have been read/resolved
+        fully_resolved_notification_states = {
+            NotificationType.collab_edit_approved,
+            NotificationType.collab_edit_revoke,
+            NotificationType.collab_revoke,
+        }
+
+        if notification_type is NotificationType.collab_request:
+            self.init_req_notification_guuid = notif.guid
+        elif notification_type is NotificationType.collab_edit_request:
+            self.edit_req_notification_guuid = notif.guid
+
+        # set necessary notification.is_resolved fields
+        elif notification_type is NotificationType.collab_approved:
+            if self.init_req_notification_guuid:
+                self._resolve_notification(self.init_req_notification_guuid)
+        elif notification_type in fully_resolved_notification_states:
+            if self.init_req_notification_guuid:
+                self._resolve_notification(self.init_req_notification_guuid)
+            if self.edit_req_notification_guuid:
+                self._resolve_notification(self.edit_req_notification_guuid)
+
+    def _resolve_notification(self, notification_guid):
```

Comment:
As a guiding principle, I think we should have modifications of an object done within that object, so this should be a classmethod in the Notification class.